[FIXED JENKINS-32765] Allow the directory that plugins are exploded into to be changed

stephenc opened the pull request:
- This really has to be controlled by either a system property or a CLI parameter as we cannot guarantee that any path
  specified in a configuration file in JENKINS_HOME will be valid on another machine where the JENKINS_HOME is mounted
- Obviously Jenkins does not currently support fully starting two Jenkins instances on the same JENKINS_HOME, so this
  fix is to assist in Disaster Recovery (or semi-higher availability) sceanarios where the backup node is being brought
  up. In such cases, the node that failed may not have correctly released all its file handles in the backing NFS share
  and thus could be blocking the recovery node from starting up.
- An additional use case is where the JENKINS_HOME is stored on a remote disk, e.g. a SAN, etc. and the user wants to
  take advantage of the faster local disk to serve the resources from both plugins and Jenkins core. Without this change
  only Jenkins core can be relocated outside of JENKINS_HOME (using the '--webroot=...' command line option).
- Ideally we would introduce this into the extras-executable-war module once the system property has had time to soak
  (although in an ideal world that module would be agnostic of Jenkins and thus it might not be appropriate there)

@reviewbybees

stephenc committed:
[FIXED JENKINS-32765] Allow the directory that plugins are exploded into to be changed

- This really has to be controlled by either a system property or a CLI parameter as we cannot guarantee that any path
  specified in a configuration file in JENKINS_HOME will be valid on another machine where the JENKINS_HOME is mounted
- Obviously Jenkins does not currently support fully starting two Jenkins instances on the same JENKINS_HOME, so this
  fix is to assist in Disaster Recovery (or semi-higher availability) sceanarios where the backup node is being brought
  up. In such cases, the node that failed may not have correctly released all its file handles in the backing NFS share
  and thus could be blocking the recovery node from starting up.
- An additional use case is where the JENKINS_HOME is stored on a remote disk, e.g. a SAN, etc. and the user wants to
  take advantage of the faster local disk to serve the resources from both plugins and Jenkins core. Without this change
  only Jenkins core can be relocated outside of JENKINS_HOME (using the '--webroot=...' command line option).
- Ideally we would introduce this into the extras-executable-war module once the system property has had time to soak
  (although in an ideal world that module would be agnostic of Jenkins and thus it might not be appropriate there)

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

amuniz commented:
:bee:

lordofthejars commented:
Should be possible to write a simple unit test of this new part of code?

oleg-nenashev commented on core/src/main/java/hudson/ClassicPluginStrategy.java:
```diff
@@ -88,6 +88,12 @@ public boolean accept(File dir,String name) {
+     * If set this should be the absolute file path to the the base directory for all exploded .hpi/.jpi plugins.
+     * If {@literal null} then plugins will be exploded into {@literal JENKINS_HOME/plugins}.
+     */
+    private static final String WORK_DIR = System.getProperty(ClassicPluginStrategy.class.getName() + ".WORK_DIR");
```
IIRC we use the cAmEl notation for system property names

lordofthejars replied:
@oleg-nenashev I am not saying you are not right :laughing: but I have seen this construction somewhere else

stephenc replied:
Normally the convention is to name the property as per the field name. This is a **static final** therefore it should be a constant thus UNDERSCORE_NAMES

stephenc replied:
Most other system properties in Jenkins are backing a run-time mutable value (i.e. they are **static** but non-final) and in those cases the field should be camelCase to reflect not being a constant. If you read the commit message and the JIRA you will see the logic as to why this cannot be runtime mutable

lordofthejars replied:
@stephenc I think that @oleg-nenashev was for the name of system property and not the variable name where the property is stored in the class.

stephenc commented:
@lordofthejars well the issue is that the unit tests run with a different plugin strategy IIRC, so this code path cannot be reached from a unit test AIUI

lordofthejars commented:
@stephenc I was thinking about doing something like:

```
cps = new ClassicPluginStrategy(....);
cps. createPluginWrapper(mock);
```

But the method is extremely big so probably quite difficult to mock everything.

lordofthejars commented:
:bee:

stephenc commented:
@reviewbybees done

reviewbybees commented:
This pull request has completed our internal processes and we now respectfully request the maintainers of this repository to consider our proposal contained within this pull request for merging.

stephenc commented:
@jenkinsci/code-reviewers

daniel-beck commented:
@stephenc You did not address the latest comment regarding naming of the system property.

stephenc commented:
@daniel-beck I did address that comment. I said that the first principle is that the system property should match the field name. As the field is actually a constant, the style in the Jenkins codebase is UPPER_CASE_WITH_UNDERSCORES. Most "constant" fields in Jenkins are actually not a constant and are run-time mutable, hence they are named using camelCaseNoSpaces. But always the property should be the same as the field name.

Therefore no change

daniel-beck commented:
@stephenc Thanks for clarifying.

stephenc committed:
[JENKINS-32765] Perhaps it makes more sense to make this a property of the PluginManager rather than the classic strategy

- Also expose via servlet context parameter of same name to allow multiple instances of Jenkins in the same servlet container to have different directories
- As we are no longer a constant, change the system property to camelCase

stephenc commented:
@reviewbybees thought about this some more, and we need to be able to support multi-tenant in the same servlet container (as I have seen people doing that... even if it is not something to recommend... we should not make it impossible)

jtnord commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -219,6 +230,14 @@ public Api getApi() {
+     * If non-null, the base directory for all exploded .hpi/.jpi plugins.
+     * @return the base directory for all exploded .hpi/.jpi plugins or {@code null} to leave this up to the strategy.
+     */
+    public File getWorkDir() {
```
@CheckForNull?

but I wonder why given this is public that you don't just return the correct workDir (rootDir) and make this `@NonNull`?

stephenc replied:
Well I considered `@CheckForNull`... but that would open the whole can of worms over **which** `@CheckForNull` the one that is in a package that is not incongruent with the Java platform policy on reserved package names, or the one claiming to be from a JSR that has never published anything - and thus is illegally squatting in the `javax` reserved namespace...

Also returning `rootDir` would change the behaviour of `ClassicPluginStrategy` which actually extracts into the same directory as the .jpi file is located and that need not necessarily be `rootDir` from the contract of a `PluginStrategy`

jtnord replied:
So make the default for `rootDir` `baseDir/plugins` ?  Any other PluginStrategy will be unaffected until they update to the new API?

FWIW the correct annotations are the deprecated edu.cs...findbugs ones :-P

stephenc committed:
[JENKINS-32765] As I am not *adding* a new import, I will annotate with `@CheckForNull`

Of course the import being used is in a package name that is reserved for artifacts produced by a JSR... and the JSR that the dependency is claiming to come from has not actually published anything... thus whoever produced the artifact has broken the JSR policy... and it is also not safe for us to actually use these annotations as we do not know what a JSR may change on the road to actually providing this... but that is a war for a different PR

stephenc committed:
[JENKINS-32765] Fix failing tests

oleg-nenashev commented:
I'm not going to block the PR due to the naming approach. BTW I cannot find any explicit recommendation about naming, and the most of system properties in the core use another naming approach.
